ml/backend/ggml/ggml.go

```diff
@@ -312,24 +320,53 @@ func New(r *os.File, params ml.BackendParams) (ml.Backend, error) {
 					return fmt.Errorf("unassigned tensor: %s", t.Name)
 				}
 
-				bts := C.malloc(C.size_t(t.Size()))
-				if bts == nil {
-					return errors.New("failed to allocate tensor buffer")
+				tts[i] = tt
+			}
+
+			bsr := bufio.NewReader(io.NewSectionReader(sr, int64(t.Offset), int64(t.Size())))
+			bts := make([]byte, 32*format.KibiByte)
+
+			var s uint64
+			for s < t.Size() {
+				n, err := io.ReadFull(bsr, bts[:min(len(bts), int(t.Size()-s))])
+				if err != nil {
+					return err
 				}
-				defer C.free(bts)
 
-				buf := unsafe.Slice((*byte)(bts), t.Size())
-				n, err := io.ReadFull(io.NewSectionReader(sr, int64(t.Offset), int64(t.Size())), buf)
-				if err != nil || n != len(buf) {
-					return errors.New("read failed")
+				for _, tt := range tts {
+					func() {
+						cbytes := C.CBytes(bts[:n])
+						defer C.free(cbytes)
+						C.ggml_backend_tensor_set(tt, cbytes, C.size_t(s), C.size_t(n))
+					}()
 				}
 
-				C.ggml_backend_tensor_set(tt, bts, 0, C.size_t(t.Size()))
-				return nil
-			})
-		}
+				s += uint64(n)
+			}
+
+			doneCount.Add(1)
+			return nil
+		})
 	}
 
+	go func() {
+		totalCount := len(meta.Tensors().Items())
+		ticker := time.NewTicker(20 * time.Millisecond)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+			case <-ticker.C:
+			}
+
+			done := doneCount.Load()
+			slog.Info("loading progress", "done", done, "total", totalCount)
```

Comment:
This sends a lot of logs. Is this the right tick rate?

Reply:
I sent out a PR to enable progress reporting to the Ollama server. I think we can just substitute in that instead of directly logging it here.

https://github.com/ollama/ollama/pull/9912